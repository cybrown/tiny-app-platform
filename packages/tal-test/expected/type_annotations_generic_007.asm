main()
  entry:
    FunctionRef       name: string_to_number_0
    DeclareLocal      name: string_to_number, mutable: false, hasInitialValue: true
    Pop               inBlock: false
    FunctionRef       name: array_map_1
    DeclareLocal      name: array_map, mutable: false, hasInitialValue: true
    Pop               inBlock: false
    Literal           "hello"
    Literal           1
    MakeArray
    FunctionRef       name: func_2
    Literal           2
    MakeArray
    Literal           0
    MakeRecord
    Local             array_map
    Call
string_to_number_0(str)
  entry:
    Literal           3
    MakeArrayForBlock count: 1
array_map_1(arr, b)
  entry:
    Literal           0
    Local             arr
    Index
    Literal           1
    MakeArray
    Literal           0
    MakeRecord
    Local             b
    Call
    Literal           1
    MakeArray
    MakeArrayForBlock count: 1
func_2(arg)
  entry:
    Local             arg
    Literal           1
    MakeArray
    Literal           0
    MakeRecord
    Local             string_to_number
    Call
