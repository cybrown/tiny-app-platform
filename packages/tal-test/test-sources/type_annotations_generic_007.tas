fun string_to_number(str: string): number {
    3
}

fun array_map<T, U>(arr: array<T>, b: (arg: T) => U): array<U> {
    [b(arr[0])]
}

array_map(["hello"], (arg) => string_to_number(arg))
