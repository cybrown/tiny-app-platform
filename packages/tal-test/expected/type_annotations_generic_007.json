[
  {
    "kind": "DeclareLocal",
    "mutable": false,
    "name": "string_to_number",
    "value": {
      "kind": "Function",
      "body": {
        "kind": "Block",
        "children": [
          3
        ]
      },
      "parameters": [
        {
          "name": "str",
          "type": {
            "kind": "named",
            "name": "string"
          }
        }
      ],
      "returnType": {
        "kind": "named",
        "name": "number"
      }
    }
  },
  {
    "kind": "DeclareLocal",
    "mutable": false,
    "name": "array_map",
    "value": {
      "kind": "Function",
      "body": {
        "kind": "Block",
        "children": [
          {
            "kind": "Array",
            "value": [
              {
                "kind": "Call",
                "args": [
                  {
                    "kind": "PositionalArgument",
                    "value": {
                      "kind": "Index",
                      "index": 0,
                      "value": {
                        "kind": "Local",
                        "name": "arr"
                      }
                    }
                  }
                ],
                "value": {
                  "kind": "Local",
                  "name": "b"
                }
              }
            ]
          }
        ]
      },
      "parameters": [
        {
          "name": "arr",
          "type": {
            "kind": "array",
            "item": {
              "kind": "named",
              "name": "T"
            }
          }
        },
        {
          "name": "b",
          "type": {
            "kind": "function",
            "parameters": [
              {
                "name": "arg",
                "type": {
                  "kind": "named",
                  "name": "T"
                }
              }
            ],
            "returnType": {
              "kind": "named",
              "name": "U"
            }
          }
        }
      ],
      "returnType": {
        "kind": "array",
        "item": {
          "kind": "named",
          "name": "U"
        }
      },
      "genericParameters": [
        {
          "name": "T"
        },
        {
          "name": "U"
        }
      ]
    }
  },
  {
    "kind": "Call",
    "args": [
      {
        "kind": "PositionalArgument",
        "value": {
          "kind": "Array",
          "value": [
            "hello"
          ]
        }
      },
      {
        "kind": "PositionalArgument",
        "value": {
          "kind": "Function",
          "body": {
            "kind": "Call",
            "args": [
              {
                "kind": "PositionalArgument",
                "value": {
                  "kind": "Local",
                  "name": "arg"
                }
              }
            ],
            "value": {
              "kind": "Local",
              "name": "string_to_number"
            }
          },
          "parameters": [
            {
              "name": "arg"
            }
          ]
        }
      }
    ],
    "value": {
      "kind": "Local",
      "name": "array_map"
    }
  }
]